editor/src/core/shared/project-contents-dependencies.ts

```diff
@@ -0,0 +1,170 @@
+import { MapLike } from 'typescript'
+import {
+  getContentsTreeFileFromString,
+  isProjectContentFile,
+  ProjectContentsTree,
+  ProjectContentTreeRoot,
+  walkContentsTree,
+  zipContentsTree,
+} from '../../components/assets'
+import { resolveModule } from '../es-modules/package-manager/module-resolution'
+import { getSavedCodeFromTextFile, getUnsavedCodeFromTextFile } from '../model/project-file-utils'
+import { addToMapOfArraysUnique } from './array-utils'
+import { isParseSuccess, isTextFile, NodeModules } from './project-file-types'
+import { emptySet } from './set-utils'
+
+export function getDirectReverseDependencies(
+  projectContents: ProjectContentTreeRoot,
+  nodeModules: NodeModules,
+): MapLike<Array<string>> {
+  let result: MapLike<Array<string>> = {}
+  walkContentsTree(projectContents, (fullPath, file) => {
+    if (isTextFile(file)) {
+      const parseResult = file.fileContents.parsed
+      if (isParseSuccess(parseResult)) {
+        for (const importSource of Object.keys(parseResult.imports)) {
+          const resolveResult = resolveModule(projectContents, nodeModules, fullPath, importSource)
+          switch (resolveResult.type) {
+            case 'RESOLVE_SUCCESS':
+              const successResult = resolveResult.success
+              switch (successResult.file.type) {
+                case 'ES_CODE_FILE':
+                  // Check if the file is part of the project.
+                  if (getContentsTreeFileFromString(projectContents, successResult.path) != null) {
+                    // Dependency, so record the connection.
+                    result = addToMapOfArraysUnique(result, successResult.path, fullPath)
+                  }
+                  break
+                case 'ES_REMOTE_DEPENDENCY_PLACEHOLDER':
+                  // Do nothing for these.
+                  break
+                default:
+                  const _exhaustiveCheck: never = successResult.file
+                  throw new Error(`Unhandled case ${JSON.stringify(successResult.file)}`)
+              }
+              break
+            case 'RESOLVE_NOT_PRESENT':
+              // Do nothing for now...
+              break
+            default:
+              const _exhaustiveCheck: never = resolveResult
+              throw new Error(`Unhandled case ${JSON.stringify(resolveResult)}`)
+          }
+        }
+      }
+    }
+  })
+  return result
+}
+
+export function getTransitiveReverseDependencies(
```

Comment:
could you also make a test for this function?

Reply:
Noted, will add some subsequently.